### DISCUSSION TITLE
两个关于内存释放 和 一个PAGComposition的问题

### DISCUSSION THREAD

**Discussion by YauLam** [Tencent/libpag#2343] [Q&A] [answered]
PAGView和PAGImageView两个类的onDetachFromWindow()中，对释放资源均有处理。
问题1: 为什么PAGView提供了freeCache()，而PAGImageView没有？

android官方提供的demo，MainActivity的onDestroy()中有对PAGView.freeCache()的调用。而PAGView源码中，在onDetachFromWindow已经调用过freeCache()了。
问题2: 开发过程中，在activity的onDestroy中，是否还有必要手动调用freeCache()？

问题3: PAGComposition和PAGLayer的组合使用，PAGComposition.Make(width, height)，创建一个空PAGComposition的时候，width和height一定要传一个指定大小吗，PAGComposition.Make(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)不可以？如果不确定动画的大小，PAGComposition.Make(width, height)里面建议传什么呢？

**Comment by kevingpqi123** [comment, marked as answer]
1、关于 freeCache，建议此类文件多看下 API 文档：
Free the cache created by the pag view immediately. Can be called to reduce memory pressure.

- (void)freeCache
Declared In
PAGView.h

如果希望立即释放一些内存，可以调用此接口，同时可以不用调用，此时内存释放则走系统的内存释放机制，在后面的某个时机释放

为啥 PAGImageView 没有，建议了解下 PAGImageView 的实现机制：https://pag.art/docs/use-pagimageview.html
这些都有文档

2、PAGComposition.Make(width, height) 指的是创建的容器的大小，就相当于一个 pag 文件的尺寸有多大，比如两个 pag 文件做切换且文件尺寸一样，传入的就是添加的 pag 文件的尺寸，如果添加了 9 个 pag文件，且希望是九宫格的效果，就是合并后的尺寸。这里不存在不确定动画大小的情况，pag 文件的尺寸本来就和 渲染 View 如PAGView 的尺寸可以不一致，这些 API 接口如果不知道怎么使用，建议参考我们提供的官方 demo ，或者参考 pag 源码中的测试用例

总结下来， API 接口使用层面的问题建议多看下 API 文档，如果不了解使用，看官方 demo 或 源码中的测试用例
